x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -194,6 +210,39 @@ void authenticateWithApiKeyIfPresent(ThreadContext ctx, ActionListener<Authentic
         }
     }
 
+    public void getRoleForApiKey(Authentication authentication, ThreadContext threadContext, CompositeRolesStore rolesStore,
+                                 FieldPermissionsCache fieldPermissionsCache, ActionListener<Role> listener) {
+        if (authentication.getAuthenticationType() != Authentication.AuthenticationType.API_KEY) {
+            throw new IllegalStateException("authentication type must be api key but is " + authentication.getAuthenticationType());
+        }
+
+        final Role preBuiltRole = threadContext.getTransient(API_KEY_ROLE_KEY);
+        if (preBuiltRole != null) {
+            listener.onResponse(preBuiltRole);
```

Comment:
I feel like it would be prudent to have an `assert` here that (somehow) checks that the pre built role belongs to this ApiKey.